File: 2 Arrays/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <math.h>

struct TextValues
{
    int letters,
        words,
        sentences;
};
typedef struct TextValues TextStatistics;

TextStatistics getCounts(char *textInput);

int main(void)
{
    string textInput = get_string("Text: ");
    TextStatistics results;
    results = getCounts(textInput);

    // Coleman-Liau Index
    // index = 0.0588 * L - 0.296 * S - 15.8
    float l = ((float)results.letters / results.words) * 100;
    float s = ((float)results.sentences / results.words) * 100;
    float res = (0.0588 * l) - (0.296 * s) - 15.8;
    int roundedRes = round(res);

    if (roundedRes < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (roundedRes >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", roundedRes);
    }
}

TextStatistics getCounts(char *textInput)
{
    TextStatistics results;

    results.letters = 0;
    results.words = 1; // counts last word
    results.sentences = 0;

    int textLength = strlen(textInput);
    for (int i = 0; i < textLength; i++)
    {
        char c = textInput[i];

        if ((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z'))
        {
            results.letters++;
        }

        if (c == ' ')
        {
            results.words++;
        }

        if (c == '.' || c == '?' || c == '!')
        {
            results.sentences += 1;
        }
    }
    return results;
}

File: 6 Python/readability.py
from cs50 import get_string


def main():
    string = get_string("Text: ")
    counts = getCounts(string)

    # Coleman-Liau Index
    # index = 0.0588 * L - 0.296 * S - 15.8
    l = (counts['letters'] / counts['words']) * 100
    s = (counts['sentences'] / counts['words']) * 100
    res = (0.0588 * l) - (0.296 * s) - 15.8
    roundedRes = round(res)

    printResult(roundedRes)


def getCounts(string):
    d = dict()
    letters = 0
    words = 1  # Counts last word
    sentences = 0

    for char in string:
        if char == ' ':
            words += 1
        elif char in ['.', '!', '?']:
            sentences += 1
        elif char.isalpha():
            letters += 1

    d['letters'] = letters
    d['words'] = words
    d['sentences'] = sentences

    return d


def printResult(roundedRes):
    if roundedRes < 1:
        print("Before Grade 1")
    elif roundedRes >= 16:
        print("Grade 16+")
    else:
        print(f"Grade {roundedRes}")


main()
